Add DR with active-passive and external storage (#3679)
theforeman/foreman-documentation: guides/common/modules/proc_preparing-for-disaster-recovery-with-active-and-passive-project-server-with-external-storage.adoc

[id="preparing-for-disaster-recovery-with-active-and-passive-project-server-with-external-storage"]
= Preparing for disaster recovery with active and passive {ProjectServer} with external storage

Create your passive {ProjectServer} as a clone of your active {ProjectServer}.
Make sure that both servers connect to the database content on your network-attached storage (NAS).

.Procedure

> . Replicate the `/var/lib/pulp` and `{postgresql-lib-dir}` directories from the active {ProjectServer} to your NAS.

Reviewer: We can reference the documentation that discusses using NAS here as that should be a prerequisite for this DR scenario.

Author: I seem to be having trouble finding something in RH docs that applies but I'll keep looking.

// Do we want to keep this generic like the virtualization scenario, or is this a good place to provide example rsync commands or something?
. Clone your active {ProjectServer}.
See xref:cloning_satellite_server[].
+
Keep the source server powered on and power off the new cloned server.
The source server remains your primary active server, while the clone server becomes the secondary passive server.
. Determine how you want to attach the database content on the NAS to your passive server:
* If you mount the storage directly by both your active and passive server, the servers will always see the same, up-to-date content.
* If you mount the storage only to your active server and replicate the storage periodically to another location, the passive server will access the data only if it takes over as the active server.
// This would be a good place to talk more about the pros and cons of these two alternatives, wouldn't it? What would be those pros and cons?

.Verification
Perform this test in an isolated staging environment:
// Does this make sense? We don't want users testing this in production, right?

. Stop the active server to mimic failure.

> // How? foreman-maintain service stop?

Reviewer: Better to turn the machine off (or halt the VM) to mimic a full outage. Alternatively, they could have a DR firewall test scenario that walls off the machine.

Author: Added.

. Start the passive server.

> // How? foreman-maintain service start?

Reviewer: I can see two scenarios:

 1. A passive server that is running all the time but inaccessible. In this case the user would simply ensure storage is attached properly and flip DNS.
 2. A passive server that is something like an image, that gets turned up, storage attached and then update is run.
 3. A machine that is spun up, restore is run with storage attached.

@evgeni Vet me here

Reviewer 2: 1. you can't count ;-)
2. when you run active/passive your nodes need to be identical and the only change you perform is "start the service" (however that is exactly defined)
3. I guess it should be something like your 1, but "inaccessible" means "services are down" (even if the machine is not accessed, running services would mean workers pick up work from the pool etc, not what you want!)

Author: I'm wondering whether all of this goes into too much detail for a testing/verification procedure. As an alternative, I could also drop line 27, which would take users directly to line 29 to tell them 'verify your passive server', however they would decide to do it. Is that just me being lazy, or would it actually make sense like this?

Reviewer: Experience tells me that these details are questions users will ask, and this guide is intended to help reduce the questions being asked by codifying our knowledge. Perhaps we say as @evgeni did in #3 - by stating inaccessible and then giving some examples.

[id="preparing-for-disaster-recovery-with-active-and-passive-project-server-with-external-storage"]
= Preparing for disaster recovery with active and passive {ProjectServer} with external storage

Create your passive {ProjectServer} as a clone of your active {ProjectServer}.

> Make sure that both servers connect to the database content on your shared storage.

Reviewer 2: What does "database content" mean here?

Author: I was trying to come up with a short way to describe `/var/lib/pulp` and `/var/lib/pgsql`. What's a good alternative here? Just "content"? Or just "data"? Or perhaps be specific like this:

```suggestion
Ensure that both servers connect to the contents of `/var/lib/pulp` and `/var/lib/pgsql` on your shared storage.
```

Reviewer 2: How about
```suggestion
Ensure that `/var/lib/pulp` and `/var/lib/pgsql` on your shared storage is available to both servers.
```

.Procedure
. Replicate the `/var/lib/pulp` and `{postgresql-lib-dir}` directories from the active {ProjectServer} to your shared storage.
. Clone your active {ProjectServer}.

> See xref:cloning_satellite_server[].

Reviewer 2: I just realized: the cloning process involves a DB-restore, which is probably not what people want.
What would probably work is: clone with `/var/lib/pgsql` being local, then turn things off and replace it with shared storage.
But that's fragile :/

Author: When you say that it's "fragile", you're actually kind of saying that that's probably not what we want either :) So what would be a good solution?

Reviewer 2: Yeah… The deployment consists of three things: packages (rpms), config (/etc) and data (/var/lib/). To get a working deployment you need all of them. As the data is quickly moving, we place it on shared storage so it can be picked up from various places. The packages and config are not, so we can keep them on "traditional" storage, but they still need to match.

The easiest way to create a matching set is a "clone" (in the general meaning of the term), but satellite-clone does more than cloning rpms/configs, it also does data.

How we can get to a good clone of the slowly moving parts only: I need to think more about it :D

Author: I created issue https://github.com/theforeman/foreman-documentation/issues/3714 where we can track this. According to https://github.com/theforeman/foreman-documentation/pull/3679#pullrequestreview-2662882118, this open item doesn't block merging.

[id="preparing-for-disaster-recovery-with-active-and-passive-project-server-with-external-storage"]
= Preparing for disaster recovery with active and passive {ProjectServer} with external storage

> Create your passive {ProjectServer} as a clone of your active {ProjectServer}.

Reviewer 2: You mean packages? Configurations? Data?
The latter is handled by the shared storage, but I agree we don't have a good solution for configurations and packages.

> Create your passive {ProjectServer} as a clone of your active {ProjectServer}.

Author: The way I'm interpreting this is:

* Data: "handled by the shared storage", which means no doc action is required (the shared storage solution is what this PR adds)
* Configurations + packages: "no good solution", which means no action is currently possible

Please let me know if I'm missing something.

[id="preparing-for-disaster-recovery-with-active-and-passive-project-server-with-external-storage"]
= Preparing for disaster recovery with active and passive {ProjectServer} with external storage

Create your passive {ProjectServer} as a clone of your active {ProjectServer}.
Make sure that both servers connect to the database content on your network-attached storage (NAS).

.Procedure
. Replicate the `/var/lib/pulp` and `{postgresql-lib-dir}` directories from the active {ProjectServer} to your NAS.
// Do we want to keep this generic like the virtualization scenario, or is this a good place to provide example rsync commands or something?
. Clone your active {ProjectServer}.
See xref:cloning_satellite_server[].
+

> Keep the source server powered on and power off the new cloned server.

Reviewer 3: `and` might imply that instruct the reader to do these in a particular sequence. I don't think you do, so perhaps this would be better:

`Keep the source server powered on. Keep the new cloned server powered off.`

> The source server remains your primary active server, while the clone server becomes the secondary passive server.

Reviewer 3: The `primary active server` phrasing implies that there is another active server that is not primary. Better go with `active primary server`.

> The source server remains your primary active server, while the clone server becomes the secondary passive server.

Reviewer 3: And same with `secondary passive server` -- better use `passive secondary server`.

. Determine how you want to attach the database content on the NAS to your passive server:

> * If you mount the storage directly by both your active and passive server, the servers will always see the same, up-to-date content.

Reviewer 3: `mount ... by` -> `mount ... on`, otherwise it seems like both of these mountings serve some greater purpose, and not just the purpose of these two servers interacting with the data.

Additionally, `mount by` is almost never used, it's usually `mount on` and less frequently `mount at`.

> * If you mount the storage only to your active server and replicate the storage periodically to another location, the passive server will access the data only if it takes over as the active server.

Reviewer 3: ```suggestion
* If you mount the storage only on your active server and replicate the storage periodically to another location, the passive server will access the data only if it takes over as the active server.
```

> * If you mount the storage only to your active server and replicate the storage periodically to another location, the passive server will access the data only if it takes over as the active server.

Reviewer 3: The `replicate the storage periodically to another location` part doesn't really make a difference AFAIU, it's just to ensure redundancy of data on the NAS. Hence, it has no bearing on when the passive server accesses the data.

However, I suppose the reason you've put it here is to help readers identify the specific plan out of the selection of plans in `21.1. Overview of recommended disaster recovery plans`, so keeping it is fine.

// This would be a good place to talk more about the pros and cons of these two alternatives, wouldn't it? What would be those pros and cons?

.Verification
Perform this test in an isolated staging environment:
// Does this make sense? We don't want users testing this in production, right?

. Mimic a full outage on the active server.
You can do this by turning the machine off, halting the virtual machine (VM) if your server runs on a VM, or isolating the machine by using a firewall.
. Start the passive server.
// How? foreman-maintain service start?
. Verify that your passive server can access the data stored on your NAS.
. Assess the functionality of the test {ProjectServer}.
See xref:Retrieving_the_Status_of_Services_{context}[].

> . Perform these verification checks regularly.

Reviewer 3: Maybe remove this step, and just add the word `regularly` to the intro of the verification procedure?

Reviewer 3: So the intro would read:
`Regularly perform this test in an isolated staging environment:`
